marketplace/netlify-build/src/app/netlify-integration.js

```diff
@@ -0,0 +1,139 @@
+  config = prepareConfig(config);
+
+  // Create build hooks for all sites.
+  const buildHookPromises = config.sites.map(siteConfig => {
+    return NetlifyClient.createBuildHook(siteConfig.netlifySiteId, accessToken);
+  });
+
+  const buildHooks = await Promise.all(buildHookPromises);
+
+  const updatedConfig = {
+    sites: config.sites.map((siteConfig, i) => {
+      return { ...siteConfig, buildHookId: buildHooks[i].id };
+    })
+  };
+
+  // Create Netlify notification hooks for all sites.
+  const netlifyHookPromises = uniqBy(updatedConfig.sites, s => s.netlifySiteId).reduce(
+    (acc, siteConfig) => {
+      const siteId = siteConfig.netlifySiteId;
+      const url = getPostPublishUrl(siteId + siteConfig.buildHookId);
+
+      const promisesForSite = NETLIFY_HOOK_EVENTS.map(event => {
+        const hook = { event, url };
+        return NetlifyClient.createNotificationHook(siteId, accessToken, hook);
+      });
+
+      return acc.concat([Promise.all(promisesForSite)]);
+    },
+    []
+  );
+
+  const netlifyHooks = await Promise.all(netlifyHookPromises);
+
+  // Merge flattened notification hook IDs to configuration.
+  updatedConfig.netlifyHookIds = netlifyHooks.reduce((acc, hooksForSite) => {
+    return acc.concat(hooksForSite.map(h => h.id));
+  }, []);
+
+  return updatedConfig;
+}
+
+export async function update({ config, accessToken }) {
+  config = prepareConfig(config);
+
+  // Remove existing hooks.
+  await removeExistingHooks({ config, accessToken });
+
+  // Remove references to removed hooks from configuration.
+  const updatedConfig = {
+    sites: config.sites.map(siteConfig => {
+      return { ...siteConfig, buildHookId: undefined };
+    })
+  };
+
+  // Proceed as in the installation step.
+  return install({ config: updatedConfig, accessToken });
+}
+
+function prepareConfig(config = {}) {
+  const sites = (config.sites || []).map(siteConfig => {
+    return { ...siteConfig, name: (siteConfig.name || '').trim() };
+  });
+
+  const netlifyHookIds = [...(config.netlifyHookIds || [])];
+
+  validateSiteConfigs(sites);
+
+  return { sites, netlifyHookIds };
+}
+
+function validateSiteConfigs(siteConfigs) {
+  // At least one site needs to be configured.
+  if (siteConfigs.length < 1) {
+    throw makeError('Provide at least one site configuration.');
+  }
+
+  // Find all site configurations with incomplete information.
+  const incomplete = siteConfigs.filter(siteConfig => {
+    return !siteConfig.netlifySiteId || siteConfig.name.length < 1;
+  });
+
+  if (incomplete.length > 0) {
+    throw makeError('Pick a Netlify site and provide a name for all configurations.');
+  }
+
+  // Display names must be unique.
+  const uniqueNames = uniqBy(siteConfigs, config => config.name);
+
+  if (uniqueNames.length !== siteConfigs.length) {
+    throw makeError('Display names must be unique.');
+  }
+}
+
+function makeError(message) {
+  const err = new Error(message);
+  err.useMessage = true;
```

Comment:
What's the reason for this?

Reply:
`notifyError` in the component checks `useMessage` on errors it catches. I use it to indicate our own errors so we know we can use `err.message` for a meaningful message. I don't want to show `TypeError: cannot call method undefined of...` if for whatever reason we fail hard. If there's no `useMessage` we use a generic error message.